Update to nrfx 3.7.0, add nRF54H20 EngB SoC and nRF54H20 DK 0.9.0

gmarull opened the pull request:
This PR adds nRF54H20 EngB SoC revision (current one) and updates nRF54H20 DK to have two revisions: 0.8.0 (current) and 0.9.0 (new and default, with the new SoC revision).

nika-nordic committed:
manifest: update hal_nordic revision to have nrfx 3.7.0

Bring latest release of nrfx: 3.7.0.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf54h20: add support for nRF54H20 EngB

nRF54H20 EngB is a re-label to the existing hardware revision for the
nRF54H20. nRF54H20 (whithout EngX) is becoming the final revision of the
SoC.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: hal_nordic: add support for nRF54H20 EngB

Add support for nRF54H20 EngB in the Nordic HAL glue code.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: introduce revision 0.8.0 (current)

The current DK is labeled as revision 0.8.0. Until now, no revisions were
needed, but a new hardware spin is coming, so we'll need to keep at least
two revisions in the near future until the final one becomes the default.
Prepare the board for that scenario. Note that the "old" nRF54H20 SoC is
now nRF54H20 EngB.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20dk: add board revision 0.9.0

Add a new revision for nRF54H20 DK: 0.9.0. This new hardware spin contains
the final nRF54H20 SoC. Treat it as the default, including twister.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented on scripts/ci/check_compliance.py:
```diff
@@ -927,6 +927,7 @@ def check_no_undef_outside_kconfig(self, kconf):
                               # Zephyr toolchain variant and therefore not
                               # visible to compliance.
         "BOARD_", # Used as regex in scripts/utils/board_v1_to_v2.py
+        "BOARD_REVISION", # Used in some Kconfig.board
```
@nordicjm @tejlmand any idea on why this is required? see https://github.com/zephyrproject-rtos/zephyr/actions/runs/10919646498/job/30307751818

nordicjm replied:
Would assume the hierarchy of how the script runs, it probably just sources board files and not zephyr's Kconfig files

gmarull replied:
Actually, this was not a fix, but https://github.com/zephyrproject-rtos/zephyr/pull/78631/commits/fe26d9e4752013681703ee50f688d8450ea87c6a (not sure if the best way, but seems to fix the problem).

gmarull committed:
scripts: ci: check_compliance: add BOARD_REVISION in Kconfig.board.v2

This is required if boards make use of such definition in their Kconfig
files. In Kconfig.board.v2, only `boards/Kconfig.v2` is loaded, but
BOARD_REVISION is part of `boards/Kconfig`, which can't be loaded in this
context.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: hal_nordic: nrfs: add support for nRF54H20 EngB

Make nrfs compatible with the EngB soc revision.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west_commands: runners: nrf: add support for nRF54H20 EngB

Make sure cpuapp/cpurad cores are detected correctly when using nRF54H20
EngB.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

adamkondraciuk committed:
drivers: timer: nrf_grtc_timer: Align Zephyr to new AUTOEN read manner

The new GRTC reading manner of the SYSCOUNTER uses hardware mechanism which
allows to keep it alive when any of CPUs is not sleeping. Otherwise
the SYSCOUNTER goes into sleep mode. Thus there is no
longer need to maintain the `CONFIG_NRF_GRTC_SLEEP_ALLOWED` symbol, however
if the user wants to have the SYSCOUNTER enabled all the time the
`CONFIG_NRF_GRTC_ALWAYS_ON` can be used instead.
The nrfx_grtc  driver no longer provides the `wakeup-read-sleep` reading
manner.
Also setting the GRTC clock source is performed by the nrfx_grtc driver so
it has been removed from the `sys_clock_driver_init()` function.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

nika-nordic commented:
@gmarull maybe we can switch to use: https://github.com/zephyrproject-rtos/hal_nordic/pull/223 and take this PR out of draft ?

gmarull commented:
@aescolar  can you check babblesim failures?

gmarull committed:
bluetooth: controller: ll_sw: nordic: align to nrfx 3.7.0

Some enums have been renamed:

- NRF_CCM_EVENT_ENDCRYPT -> NRF_CCM_EVENT_END
- NRF_CCM_TASK_CRYPT -> NRF_CCM_TASK_START

Also adjust some TX power level settings (26/28dBm not always available
depending on the SoC revision).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -1042,6 +1039,14 @@
     #include <nrfx_config_nrf5340_application.h>
 #elif defined(NRF5340_XXAA_NETWORK)
     #include <nrfx_config_nrf5340_network.h>
+#elif defined(NRF54H20_ENGB_XXAA) && defined(NRF_APPLICATION)
+    #include <nrfx_config_nrf54h20_engb_application.h>
+#elif defined(NRF54H20_ENGB_XXAA) && defined(NRF_RADIOCORE)
+    #include <nrfx_config_nrf54h20_engb_radiocore.h>
+#elif defined(NRF54H20_ENGB_XXAA) && defined(NRF_PPR)
+    #include <nrfx_config_nrf54h20_engb_ppr.h>
+#elif defined(NRF54H20_ENGB_XXAA) && defined(NRF_FLPR)
+    #include <nrfx_config_nrf54h20_engb_flpr.h>
```
Do we actually need a second copy of these config files? Why not just use:
```
#elif (defined(NRF54H20_XXAA) || defined(NRF54H20_ENGB_XXAA)) && defined(NRF_APPLICATION)
    #include <nrfx_config_nrf54h20_application.h>
```
and so on?

nika-nordic replied:
@gmarull yes, nrfx_config_* for these two targets are identical

gmarull replied:
adjusted

anangl commented on modules/hal_nordic/nrfx/nrfx_config_nrf54h20_application.h:
```diff
@@ -312,7 +313,7 @@
  */
 #ifndef NRFX_GPIOTE_CONFIG_NUM_OF_EVT_HANDLERS
-#define NRFX_GPIOTE_CONFIG_NUM_OF_EVT_HANDLERS 1
+#define NRFX_GPIOTE_CONFIG_NUM_OF_EVT_HANDLERS 2
```
This is unwanted change. Similarly for other cores.

anangl commented on modules/hal_nordic/nrfx/nrfx_config_nrf54h20_radiocore.h:
```diff
@@ -11,6 +11,10 @@
 #error "This file should not be included directly. Include nrfx_config.h instead."
 #endif
 
+#ifndef NRFX_RTC0_ENABLED
+#define NRFX_RTC0_ENABLED 1
+#endif
+
```
This is also unwanted. It should be fixed in nrfx (@nika-nordic).

gmarull committed:
modules: hal_nordic: align PDM configuration

Introduce instance 0, PDM0, following nrfx 3.7.0 update.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

aescolar reviewed (commented):
```diff
diff --git a/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h b/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h
index b0d3948aefb..8597ea7d586 100644
--- a/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h
+++ b/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h
@@ -633,9 +633,16 @@ static inline uint32_t hal_radio_tx_power_value(int8_t tx_power_lvl)
                return RADIO_TXPOWER_TXPOWER_Neg20dBm;
        }
 
+#if defined(RADIO_TXPOWER_TXPOWER_Neg26dBm)
        if (tx_power_lvl >= -26) {
                return RADIO_TXPOWER_TXPOWER_Neg26dBm;
        }
+#endif
+#if defined(RADIO_TXPOWER_TXPOWER_Neg28dBm)
+       if (tx_power_lvl >= -28) {
+               return RADIO_TXPOWER_TXPOWER_Neg28dBm;
+       }
+#endif
 
        if (tx_power_lvl >= -40) {
                return RADIO_TXPOWER_TXPOWER_Neg40dBm;
```

gmarull committed:
modules: hal_nordic: add nrf54l system sources to bsim build

Include system_nrf54l.c file when building for bsim.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

aescolar commented:
And
```diff
diff --git a/modules/hal_nordic/nrfx/CMakeLists.txt b/modules/hal_nordic/nrfx/CMakeLists.txt
index fde999b556a..595da6df111 100644
--- a/modules/hal_nordic/nrfx/CMakeLists.txt
+++ b/modules/hal_nordic/nrfx/CMakeLists.txt
@@ -103,7 +103,7 @@ zephyr_library_sources_ifdef(CONFIG_SOC_NRF52840       ${MDK_DIR}/system_nrf5284
zephyr_library_sources_ifdef(CONFIG_SOC_NRF5340_CPUAPP ${MDK_DIR}/system_nrf5340_application.c)
zephyr_library_sources_ifdef(CONFIG_SOC_NRF5340_CPUNET ${MDK_DIR}/system_nrf5340_network.c)
zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_NRF54HX ${MDK_DIR}/system_nrf54h.c)
-zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_NRF54LX ${MDK_DIR}/system_nrf54l.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_COMPATIBLE_NRF54LX ${MDK_DIR}/system_nrf54l.c)
zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_NRF91X  ${MDK_DIR}/system_nrf91.c)
zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_NRF92X  ${MDK_DIR}/system_nrf92.c)
```

aescolar commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -95,7 +103,9 @@ zephyr_library_sources_ifdef(CONFIG_SOC_NRF52840       ${MDK_DIR}/system_nrf5284
 zephyr_library_sources_ifdef(CONFIG_SOC_NRF5340_CPUNET ${MDK_DIR}/system_nrf5340_network.c)
 zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_NRF54HX ${MDK_DIR}/system_nrf54h.c)
-zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_NRF54LX ${MDK_DIR}/system_nrf54l.c)
+if(CONFIG_SOC_SERIES_NRF54LX OR CONFIG_SOC_SERIES_BSIM_NRF54LX)
```
CONFIG_SOC_COMPATIBLE_NRF54LX means that :)

gmarull replied:
I think we should avoid intermediate symbols here.

aescolar replied:
We are already using those COMPATIBLE ones in many places. They were introduced for this kind of cases.

gmarull committed:
drivers: adc: nrfx_saadc: use CONFIG_NRF_PLATFORM_HALTIUM

So that any Haltium based SoC uses the same settings, including nRF54H20
EngB.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: lib: cpp: cxx: ignore nRF54H20 DK 0.8.0

nRF54H HAL is not compatible with C++98, add new board revisions to the
list so that they are skipped.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

anangl commented on drivers/timer/nrf_grtc_timer.c:
```diff
@@ -513,6 +507,9 @@ static int sys_clock_driver_init(void)
 	nrfx_grtc_clock_source_set(NRF_GRTC_CLKSEL_LFXO);
 #endif
 
+#if IS_ENABLED(CONFIG_NRF_GRTC_ALWAYS_ON)
```
```suggestion
#if defined(CONFIG_NRF_GRTC_ALWAYS_ON)
```

anangl commented on drivers/timer/Kconfig.nrf_grtc:
```diff
@@ -15,11 +15,10 @@ menuconfig NRF_GRTC_TIMER
-	def_bool y
-	depends on POWEROFF
+config NRF_GRTC_ALWAYS_ON
+	def_bool n
```
Why not just `bool`? `n` is anyway the default value.

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -19,7 +19,7 @@ LOG_MODULE_REGISTER(adc_nrfx_saadc);
 #if (NRF_SAADC_HAS_AIN_AS_PIN)
 
-#if defined(CONFIG_SOC_NRF54H20) || defined(CONFIG_SOC_NRF9280)
+#if defined(CONFIG_SOC_NRF54H20) || defined(CONFIG_SOC_NRF54H20_ENGB) || defined(CONFIG_SOC_NRF9280)
```
```suggestion
#if defined(CONFIG_NRF_PLATFORM_HALTIUM)
```

nika-nordic commented:
https://github.com/zephyrproject-rtos/hal_nordic/pull/223 merged, please update SHA and rebase
